src/axom/primal/geometry/Polygon.hpp

```diff
@@ -22,63 +22,192 @@ namespace axom
+ *         if array type is dynamic.
  * \note The polygon vertices should be ordered in a counter clockwise
  *       orientation with respect to the polygon's desired normal vector
  */
-template <typename T, int NDIMS>
+template <typename T,
+          int NDIMS,
+          PolygonArray ARRAY_TYPE = PolygonArray::Dynamic,
+          int MAX_VERTS = DEFAULT_MAX_NUM_VERTICES>
 class Polygon
 {
 public:
   using PointType = Point<T, NDIMS>;
   using VectorType = Vector<T, NDIMS>;
 
+  // axom::Array for dynamic array type, axom::StackArray for static array type
+  using ArrayType = std::conditional_t<ARRAY_TYPE == PolygonArray::Dynamic,
+                                       axom::Array<PointType>,
+                                       axom::StackArray<PointType, MAX_VERTS>>;
+
 public:
-  /// Default constructor for an empty polygon
-  Polygon() { }
+  /// Default constructor for an empty polygon (dynamic array specialization).
+  /// Specializations are necessary to remove __host__ __device__ warning for
+  /// axom::Array usage with the dynamic array type.
+  template <PolygonArray P_ARRAY_TYPE = ARRAY_TYPE,
+            std::enable_if_t<P_ARRAY_TYPE == PolygonArray::Dynamic, int> = 0>
+  Polygon()
+  { }
+
+  /// Default constructor for an empty polygon (static array specialization)
+  template <PolygonArray P_ARRAY_TYPE = ARRAY_TYPE,
+            std::enable_if_t<P_ARRAY_TYPE == PolygonArray::Static, int> = 0>
+  AXOM_HOST_DEVICE Polygon()
+  { }
 
   /*!
    * \brief Constructor for an empty polygon that reserves space for
-   *  the given number of vertices
+   *  the given number of vertices. Only available for dynamic arrays.
    *
    * \param [in] numExpectedVerts number of vertices for which to reserve space
    * \pre numExpectedVerts is not negative
    *
    */
+  template <PolygonArray P_ARRAY_TYPE = ARRAY_TYPE,
+            std::enable_if_t<P_ARRAY_TYPE == PolygonArray::Dynamic, int> = 0>
   explicit Polygon(int numExpectedVerts)
   {
     SLIC_ASSERT(numExpectedVerts >= 0);
     m_vertices.reserve(numExpectedVerts);
   }
 
   /// \brief Constructor for a polygon with the given vertices
-  Polygon(const axom::Array<PointType>& vertices) { m_vertices = vertices; }
+  ///        (dynamic array specialization)
+  template <PolygonArray P_ARRAY_TYPE = ARRAY_TYPE,
+            std::enable_if_t<P_ARRAY_TYPE == PolygonArray::Dynamic, int> = 0>
+  explicit Polygon(const axom::Array<PointType>& vertices)
+  {
+    m_vertices = vertices;
+    m_num_vertices = m_vertices.size();
+  }
+
+  /// \brief Constructor for a polygon with the given vertices
+  ///        (static array specialization)
+  template <PolygonArray P_ARRAY_TYPE = ARRAY_TYPE,
+            std::enable_if_t<P_ARRAY_TYPE == PolygonArray::Static, int> = 0>
+  explicit Polygon(const axom::Array<PointType>& vertices)
+  {
+    SLIC_ASSERT(static_cast<int>(vertices.size()) <= MAX_VERTS);
+
+    for(int i = 0; i < vertices.size(); i++)
```

Comment:
For consistency, should you use the iterator syntax, like you do below at line 150, instead of a for-loop? Here, and elsewhere.